### PR TITLE
Adding electron input streams for Ferminet

### PR DESCRIPTION
**Pull request by shaipranesh2**
# Pull Request Template

## Description
This is a one of several PRs aimed at implementing Ferminet. This PR aims to implement the input stream function
which calculate the input needed for the model (one-electron and two-electron stream).

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: requirements/env_common.yml [hunk 1]
```diff
@@ -26,3 +26,4 @@ dependencies:
     - xgboost
     - git+https://github.com/samoturk/mol2vec
     - tb-nightly # @arunppsg shift to tensorboard stable version once tb 2.9.1 is released
+    - pyscf
```
Comment (peastman) [inline requirements/env_common.yml:29]: What is this needed for?  Adding a new dependency to DeepChem should be done only when absolutely necessary and needs to be carefully considered first.

Comment (shaipranesh2) [reply requirements/env_common.yml:29]: Yeah I wanted to discuss this over in the developer calls as well. I thought this was necessary as ferminet needs a baseline solution of Hartree-Fock for pre-training which can be obtained from pySCF.

Comment (tonydavis629) [reply requirements/env_common.yml:29]: We can add it as a soft dependency, no? 

I think it may be better to use the pyscf conda channel to install over installing through pip.

File: deepchem/models/torch_models/ferminet.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    no_electrons = []
+    nucleons = []
+
+    for i in self.nucleon_coordinates:
+      mol=Chem.MolFromSmiles(i[0])
+      for j in mol.GetAtoms():
+        nucleons.append(j.GetAtomicNum()-j.GetFormalCharge())
+      nucleons.append(i[1])
+
+    electron_no = np.array(no_electrons)
```
Comment (tonydavis629) [inline deepchem/models/torch_models/ferminet.py:72]: Why not just assigned electron_no as an empty array?

File: deepchem/models/tests/test_ferminet.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  h2_molecule = [['[H]', [0, 0, 0]], ['[H]', [0, 0, 0.748]]]
+  molecule = Ferminet(h2_molecule, seed_no=0, batch_number=1)
+  molecule.prepare_input_stream()
+
+  assert (molecule.one_electron_vector == np.array(
+      [[[0.035281046919353284, 0.008003144167344467, 0.019574759682114785],
+        [0.035281046919353284, 0.008003144167344467, -0.7284252403178852]],
+       [[0.044817863984029156, 0.03735115980299935, 0.7284544424024718],
+        [0.044817863984029156, 0.03735115980299935,
+         -0.019545557597528185]]])).all()
```
Comment (peastman) [inline deepchem/models/tests/test_ferminet.py:26]: It would be better to use `allclose()` for these assertions.  Very tiny changes, even just running on a different processor, can affect values at this level of precision.

File: deepchem/models/torch_models/ferminet.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  def __init__(self,
+               nucleon_coordinates: List[List],
+               seed_no: Optional[int] = None,
+               batch_number: int = 10):
+    """
+    Parameters:
+    -----------
+    nucleon_coordinates:  List[List]
+      A dictionary containing nucleon coordinates as the values with the keys as the element's symbol.
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:42]: It's a list, not a dict.

File: deepchem/models/torch_models/ferminet.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    This method is based on the following paper:
+
+    Spencer, James S., et al. Better, Faster Fermionic Neural Networks. arXiv:2011.07125,
+    arXiv, 13 Nov. 2020. arXiv.org, http://arxiv.org/abs/2011.07125.
+    """
+
+  def __init__(self,
+               nucleon_coordinates: List[List],
+               seed_no: Optional[int] = None,
+               batch_number: int = 10):
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:37]: ElectronSampler calls these arguments `seed` and `batch_no`.  We should make the two classes consistent.

File: deepchem/models/torch_models/ferminet.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    two_electron_vector: numpy.ndarray
+      The two-electron input stream containing the distance vector between all the electrons coordinates.
+    two_electron_distance: numpy.ndarray
+    """
+
+    no_electrons = []
+    nucleons = []
+
+    for i in self.nucleon_coordinates:
+      mol = Chem.MolFromSmiles(i[0])
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:78]: Are they actually SMILES strings, or just element symbols as the docstring says?

Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py]: it should be just the elements symbol. Now, I have made changes such that works with element symbols

File: deepchem/models/torch_models/ferminet.py (hunk 5 above)
Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py:37]: I didn't quite understand the "class consistency" here 😅. Did u mean like I should remove the seed_no and batch_number here in ferminet class? If so, then how will the user set the seed and batch number then?

Comment (tonydavis629) [reply deepchem/models/torch_models/ferminet.py:38]: I think he means the name of the arguments should be changed from `seed_no` to `seed` and `batch_number` to `batch_no`.

File: deepchem/models/torch_models/ferminet.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    -----------
+    nucleon_coordinates:  List[List]
+      A list containing nucleon coordinates as the values with the keys as the element's symbol.
+    seed_no: int, optional (default None)
+      Random seed to use for electron initialization.
+    batch_number: int, optional (default 10)
+      Number of batches of the electron's positions to be initialized.
+
+    """
+    super(Ferminet, self).__init__()
```
Comment (tonydavis629) [inline deepchem/models/torch_models/ferminet.py:50]: There is no super class for Ferminet to call

File: deepchem/models/torch_models/ferminet.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    no_electrons = []
+    nucleons = []
+    self.charge: List = []
+
+    for i in self.nucleon_coordinates:
+      mol = Chem.MolFromSmiles('[' + i[0] + ']')
+      for j in mol.GetAtoms():
+        self.charge.append(j.GetAtomicNum())
+        no_electrons.append([j.GetAtomicNum() - j.GetFormalCharge()])
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:84]: This is a very expensive way to look up the atomic number for a symbol.  A much cheaper way is

```python
table = Chem.GetPeriodicTable()
...
num = table.GetAtomicNumber(i[0])
```

The formal charge will always be 0, because you don't specify a formal charge.  But should there be an argument somewhere for the user to specify the total charge of the system?

File: deepchem/models/torch_models/ferminet.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      nucleons.append(i[1])
+
+    self.electron_no: np.ndarray = np.array(no_electrons)
+    self.nucleon_pos: np.ndarray = np.array(nucleons)
+
+    spin = np.sum(self.electron_no) % 2
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:90]: Is this meant to be the total spin of the system, or is it something specific to the model?  Normally the user specifies the total spin, since the ground state isn't always the one with the smallest spin.

Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py:90]: yeah, this is meant for a single atom, have changed it now to get the user's input

File: deepchem/models/torch_models/ferminet.py (hunk 8 above)
Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py:84]: Have changed it now. and for specifying the formal charge, I have allowed the user to input it with the element string itself, like 'H1+' or 'H1-', where the end of the string defines the charge type and the number before it defines the magnitude. Is this fine?

Comment (peastman) [reply deepchem/models/torch_models/ferminet.py]: Usually the way quantum chemistry programs handle this is that you specify the elements, the total charge, and the spin multiplicity.  See for example https://psicode.org/psi4manual/master/tutorial.html.  By specifying formal charges, you imply that you know where all the electrons are going to end up.  But generally you don't know, and it's the program's job to figure out where they'll go.

Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py]: Since at the start we sample electrons around each atom, in case of a formal charge, an atom is randomly selected for the charge?(I am unsure on how the program will figure out at the start)

**Comment by shaipranesh2** [comment]
@peastman @tonydavis629 , for the electronegativity data, I found this csv (https://github.com/lewisacidic/scikit-chem/blob/cbc9765e63cc78e771d6c31b5c708dfaf2a19df5/skchem/resource/atom_data.csv), containing electronegative and a few other properties. So should I go ahead with this data?

**Comment by peastman** [comment]
Sure, it seems worth a try.  Wikipedia also has [several different electronegativity scales](https://en.wikipedia.org/wiki/Electronegativities_of_the_elements_(data_page)).  I don't think it matters which one we use.  We just need some reasonable way of ordering the elements.

**Comment by shaipranesh2** [comment]
@tonydavis629 @peastman i guess the CI errors are not related to this code. But is this way of calculating the electronegativity fine?

**Comment by shaipranesh2** [comment]
@tonydavis629 @peastman the errors are not related to the code here I guess. And is this way of using electronegativity fine?

File: deepchem/models/torch_models/ferminet.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+      seed: Optional[int] = None,
+      batch_no: int = 10,
+  ):
+    """
+    Parameters:
+    -----------
+    nucleon_coordinates:  List[List]
+      A list containing nucleon coordinates as the values with the keys as the element's symbol.
+    spin: int
+      The total spin of the molecule system.
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:56]: You probably mean spin multiplicity?  Spin isn't an integer.

File: deepchem/models/torch_models/ferminet.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+      elif len(
+          nucleons
+      ) == 2:  # for a diatomic molecule, the most electronegative atom will get the anionic charge
+        electro_neg = electro_neg[electro_neg[:, 1].argsort()]
+        if self.ion_charge > 0:
+          pos = electro_neg[0][0]
+        else:
+          pos = electro_neg[-1][0]
+        self.electron_no[int(pos)][0] -= self.ion_charge
+      else:  # for a multiatomic molecule, the atom's electronegativity is averaged out with the weight sum of neighbouring atom's electronegativity and their interatomic distance
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:125]: I'm not familiar with this calculation.  But I'm also not a chemist.  What's the reason for it?

File: deepchem/models/torch_models/ferminet.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            self.electron_no[int(electro_neg[pos][0])][0] -= per_atom_charge
+            pos += increment
+          self.electron_no[int(electro_neg[pos -
+                                           increment][0])][0] += extra_charge
+        else:
+          if self.ion_charge > 0:
+            pos = 0
+          else:
+            pos = -1
+          self.electron_no[int(electro_neg[pos][0])][0] -= self.ion_charge
```
Comment (peastman) [inline deepchem/models/torch_models/ferminet.py:157]: Adding multiple extra electrons to a single atom probably won't work well.  Highly charged atoms are unusual and tend to have high energies.  For example, if the total charge is -2 you probably should assign -1 to each of the two most electronegative atoms, not -2 to a single atom.

File: deepchem/models/torch_models/ferminet.py (hunk 11 above)
Comment (peastman) [reply deepchem/models/torch_models/ferminet.py:116]: Remember to remove the comment, since it doesn't apply anymore.

File: deepchem/models/torch_models/ferminet.py (hunk 10 above)
Comment (peastman) [reply deepchem/models/torch_models/ferminet.py:56]: Remember to fix this.

**Review by peastman — APPROVED** [review]
